qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -61,42 +61,37 @@ class LinearFunction(Gate):
     `Online at umich.edu. <https://web.eecs.umich.edu/~imarkov/pubs/jour/qic08-cnot.pdf>`_
     """
 
-    def __init__(
-        self,
-        linear: list[list[int]] | np.ndarray | QuantumCircuit,
-        validate_input: bool | None = False,
-    ) -> None:
+    def __init__(self, linear, validate_input=False):
         """Create a new linear function.
 
         Args:
-            linear (list[list] or ndarray[bool] or QuantumCircuit):
-                either an n x n matrix, describing the linear function,
-                or a quantum circuit composed of linear gates only
-                (currently supported gates are CX and SWAP).
+            linear (list[list] or ndarray[bool] or QuantumCircuit or LinearFunction
+                or PermutationGate or Clifford): data from which a linear function
+                can be constructed. It can be either a nxn matrix (describing the
+                linear transformation), a permutation (which is a special case of
+                a linear function), another linear function, a clifford (when it
+                corresponds to a linear function), or a quantum circuit composed of
+                linear gates (CX and SWAP) and other objects described above, including
+                nested subcircuits.
 
             validate_input: if True, performs more expensive input validation checks,
                 such as checking that a given n x n matrix is invertible.
 
         Raises:
             CircuitError: if the input is invalid:
-                either a matrix is non {square, invertible},
-                or a quantum circuit contains non-linear gates.
-
+                either the input matrix is not square or not invertible,
+                or the input quantum circuit contains non-linear objects
+                (for example, a Hadamard gate, or a Clifford that does
+                not correspond to a linear function).
         """
-        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
-            raise CircuitError(
-                "A linear function must be represented either by a list, "
-                "a numpy array, or a quantum circuit with linear gates."
-            )
 
-        if isinstance(linear, QuantumCircuit):
-            # The following function will raise a CircuitError if there are nonlinear gates.
-            original_circuit = linear
-            linear = _linear_quantum_circuit_to_mat(linear)
+        # pylint: disable=cyclic-import
+        from qiskit.circuit.library import PermutationGate
```

Comment:
I think you can avoid the cyclical import issue here if you use:

 `from qiskit.circuit.library.generalized_gates.permutation import PermutationGate`
 
 then you can move it to the top. Or just use a relative import since they're in the same directory `from .permutation import PermutationGate`, but I generally prefer absolute paths (an old habit from Python 2 where relative imports were unreliable).

Reply:
Thanks, this is done in [5a1cf9f](https://github.com/Qiskit/qiskit-terra/pull/10053/commits/5a1cf9fd81e56b51219e0cd68a06c65341844274). Though, if we will later want to add method `from_linear_function` to `PermutationGate`, then we would not be able to also top-level-import `LinearFunction` in `permutation_gate.py`, or could we?